src/librustc/mir/mod.rs

```diff
@@ -2866,7 +2866,7 @@ impl<'tcx> TypeFoldable<'tcx> for PlaceElem<'tcx> {
             Deref => Deref,
             Field(f, ty) => Field(*f, ty.fold_with(folder)),
             Index(v) => Index(v.fold_with(folder)),
-            elem => elem.clone(),
+            elem => *elem,
```

Comment:
Actually, the way this is structured is probably a bad idea and should be exhaustive, hmpf.

Reply:
````diff
--- a/src/librustc/mir/mod.rs
+++ b/src/librustc/mir/mod.rs
@@ -2466,7 +2466,13 @@ impl<'tcx> TypeFoldable<'tcx> for UserTypeProjection {
                 Deref => Deref,
                 Field(f, ()) => Field(*f, ()),
                 Index(()) => Index(()),
-                elem => *elem,
+                Downcast(symbol, variantidx) => Downcast(*symbol, *variantidx),
+                ConstantIndex { offset, min_length, from_end } => {
+                    ConstantIndex { offset: *offset, min_length: *min_length, from_end: *from_end }
+                }
+                Subslice { from, to, from_end } => {
+                    Subslice { from: *from, to: *to, from_end: *from_end }
+                }
             })
             .collect();

````